server/src/main/java/org/elasticsearch/index/mapper/AnnotatedTextFieldMapper.java

```diff
@@ -0,0 +1,937 @@
+    // needs plain text format for highlighting but marked-up format for token discovery.
+    // The class takes markedup format field values and returns plain text versions.
+    // When asked to tokenize plain-text versions by the highlighter it tokenizes the
+    // original markup form in order to inject annotations.
+    public static final class AnnotatedHighlighterAnalyzer extends AnalyzerWrapper {
+        Map<String, AnnotatedText> plainToMarkedupForm = new HashMap<>();
+        private Analyzer delegate;
+        public AnnotatedHighlighterAnalyzer(List<Object> markedupFieldValues, Analyzer delegate){
+            super(delegate.getReuseStrategy());
+            this.delegate = delegate;
+            for (Object object : markedupFieldValues) {
+                String markedUpText = object.toString();
+                AnnotatedText at = new AnnotatedText(markedUpText);
+                plainToMarkedupForm.put(at.textMinusMarkup, at);
+            }
+        }
+        
+        public List<Object> getPlainTextValuesForHighlighter(){
+            return new ArrayList<>(plainToMarkedupForm.keySet());
+        }
+        
+        @Override
+        public Analyzer getWrappedAnalyzer(String fieldName) {
+          return delegate;
+        }   
+        
+        @Override
+        protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
+            if(components instanceof AnnotatedHighlighterTokenStreamComponents){
+                // already wrapped.
+                return components;
+            }
+            AnnotationsInjector injector = new AnnotationsInjector(components.getTokenStream());
+            return new AnnotatedHighlighterTokenStreamComponents(components.getTokenizer(), injector, this.plainToMarkedupForm);
+        }        
+    }
+    private static final class AnnotatedHighlighterTokenStreamComponents extends TokenStreamComponents{
+
+        private AnnotationsInjector annotationsInjector;
+        private Map<String, AnnotatedText> plainToMarkedUpForm;
+
+        public AnnotatedHighlighterTokenStreamComponents(Tokenizer source, AnnotationsInjector annotationsFilter,
+                Map<String, AnnotatedText> plainToMarkedupForm) {
+            super(source, annotationsFilter);
+            this.annotationsInjector = annotationsFilter;
+            this.plainToMarkedUpForm = plainToMarkedupForm;
+        }
+
+        @Override
+        protected void setReader(Reader reader) {
+            String plainText = readToString(reader);
+            AnnotatedText at = plainToMarkedUpForm.get(plainText);
+            if(at == null){
+                throw new IllegalStateException("Failed to find pre-parsed text: [" +plainText+"]");
+            }
+            annotationsInjector.setAnnotations(at);
+            super.setReader(new StringReader(at.textMinusMarkup));  
+        }
+               
```

Comment:
Can you do this instead in AnnotatedHighlighterAnalyzer.wrapReader()?  That would solve your issues with further wrapping, I think.

Reply:
I use wrapReader in the main AnnotationAnalyzerWrapper which is now a wrappable Analyzer (courtesy of some ThreadLocal ickyness).
The line you've highlighted is from a special Analyzer designed specifically for use with highlighting annotated fields. The challenge there is a different one - the UnifiedHighlighter calls an Analyzer with fragments of  text (eg sentences) and the AnnotatedHighlighterAnalyzer is responsible for mapping the plain-text string sentence presented by the highlighter to it's marked-up counterpart so that it can produce the Annotation tokens.